File: Application/src/main/java/com/shareconnect/ProfileManager.java
package com.shareconnect;

import android.content.Context;
import android.content.SharedPreferences;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class ProfileManager {
    private static final String PREFS_NAME = "MeTubeSharePrefs";
    private static final String KEY_PROFILES = "profiles";
    private static final String KEY_DEFAULT_PROFILE = "default_profile";

    private SharedPreferences sharedPreferences;
    private Gson gson;

    public ProfileManager(Context context) {
        sharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        gson = new Gson();
    }

    public List<ServerProfile> getProfiles() {
        String profilesJson = sharedPreferences.getString(KEY_PROFILES, null);
        if (profilesJson != null) {
            Type listType = new TypeToken<List<ServerProfile>>(){}.getType();
            return gson.fromJson(profilesJson, listType);
        }
        return new ArrayList<>();
    }

    public void saveProfiles(List<ServerProfile> profiles) {
        String profilesJson = gson.toJson(profiles);
        sharedPreferences.edit().putString(KEY_PROFILES, profilesJson).apply();
    }

    public ServerProfile getDefaultProfile() {
        String defaultProfileId = sharedPreferences.getString(KEY_DEFAULT_PROFILE, null);
        if (defaultProfileId != null) {
            List<ServerProfile> profiles = getProfiles();
            for (ServerProfile profile : profiles) {
                if (profile.getId().equals(defaultProfileId)) {
                    return profile;
                }
            }
        }
        // If no default profile is set, return the first profile if available
        List<ServerProfile> profiles = getProfiles();
        if (!profiles.isEmpty()) {
            return profiles.get(0);
        }
        return null;
    }

    public void setDefaultProfile(ServerProfile profile) {
        sharedPreferences.edit().putString(KEY_DEFAULT_PROFILE, profile.getId()).apply();
    }

    public void addProfile(ServerProfile profile) {
        if (profile.getId() == null || profile.getId().isEmpty()) {
            profile.setId(UUID.randomUUID().toString());
        }
        
        // Set default service type if not set
        if (profile.getServiceType() == null || profile.getServiceType().isEmpty()) {
            profile.setServiceType(ServerProfile.TYPE_METUBE);
        }
        
        List<ServerProfile> profiles = getProfiles();
        profiles.add(profile);
        saveProfiles(profiles);
    }

    public void updateProfile(ServerProfile profile) {
        List<ServerProfile> profiles = getProfiles();
        for (int i = 0; i < profiles.size(); i++) {
            if (profiles.get(i).getId().equals(profile.getId())) {
                profiles.set(i, profile);
                break;
            }
        }
        saveProfiles(profiles);
    }

    public void deleteProfile(ServerProfile profile) {
        List<ServerProfile> profiles = getProfiles();
        profiles.remove(profile);
        saveProfiles(profiles);
        
        // If we're deleting the default profile, clear the default setting
        String defaultProfileId = sharedPreferences.getString(KEY_DEFAULT_PROFILE, null);
        if (defaultProfileId != null && defaultProfileId.equals(profile.getId())) {
            sharedPreferences.edit().remove(KEY_DEFAULT_PROFILE).apply();
        }
    }

    public boolean hasProfiles() {
        return !getProfiles().isEmpty();
    }
    
    /**
     * Get profiles filtered by service type
     */
    public List<ServerProfile> getProfilesByServiceType(String serviceType) {
        List<ServerProfile> allProfiles = getProfiles();
        List<ServerProfile> filteredProfiles = new ArrayList<>();
        
        for (ServerProfile profile : allProfiles) {
            if (serviceType.equals(profile.getServiceType())) {
                filteredProfiles.add(profile);
            }
        }
        
        return filteredProfiles;
    }
    
    /**
     * Get all unique service types from existing profiles
     */
    public List<String> getAllServiceTypes() {
        List<ServerProfile> allProfiles = getProfiles();
        List<String> serviceTypes = new ArrayList<>();
        
        for (ServerProfile profile : allProfiles) {
            String serviceType = profile.getServiceType();
            if (serviceType != null && !serviceTypes.contains(serviceType)) {
                serviceTypes.add(serviceType);
            }
        }
        
        return serviceTypes;
    }
    
    /**
     * Get all torrent client types from existing profiles
     */
    public List<String> getAllTorrentClientTypes() {
        List<ServerProfile> allProfiles = getProfiles();
        List<String> clientTypes = new ArrayList<>();
        
        for (ServerProfile profile : allProfiles) {
            if (ServerProfile.TYPE_TORRENT.equals(profile.getServiceType())) {
                String clientType = profile.getTorrentClientType();
                if (clientType != null && !clientTypes.contains(clientType)) {
                    clientTypes.add(clientType);
                }
            }
        }
        
        return clientTypes;
    }
}

File: Application/src/main/kotlin/com/shareconnect/ProfileManager.kt
package com.shareconnect

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.shareconnect.database.ServerProfileRepository
import java.util.UUID

class ProfileManager(private val context: Context) {
    private val repository: ServerProfileRepository
    private val sharedPreferences: SharedPreferences
    private val gson: Gson

    init {
        repository = ServerProfileRepository(context)
        sharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        gson = Gson()

        // Migrate from SharedPreferences to Room if needed
        migrateIfNeeded()
    }

    private fun migrateIfNeeded() {
        if (!sharedPreferences.getBoolean("migrated_to_room", false)) {
            val profilesJson = sharedPreferences.getString(KEY_PROFILES, null)
            if (profilesJson != null) {
                val listType = object : TypeToken<List<ServerProfile>>() {}.type
                val oldProfiles: List<ServerProfile> = gson.fromJson(profilesJson, listType)

                // Migrate each profile to Room database
                oldProfiles.forEach { profile ->
                    if (profile.id.isNullOrEmpty()) {
                        profile.id = UUID.randomUUID().toString()
                    }
                    repository.addProfile(profile)
                }

                // Mark as migrated
                sharedPreferences.edit()
                    .putBoolean("migrated_to_room", true)
                    .remove(KEY_PROFILES) // Clean up old data
                    .apply()
            } else {
                // No old profiles, just mark as migrated
                sharedPreferences.edit()
                    .putBoolean("migrated_to_room", true)
                    .apply()
            }
        }
    }

    val profiles: List<ServerProfile>
        get() = repository.getAllProfiles()

    fun saveProfiles(profiles: List<ServerProfile>) {
        // This method is kept for backward compatibility but now uses Room
        // Clear all existing profiles and add new ones
        // Note: In a production app, this would be more sophisticated
        profiles.forEach { profile ->
            if (repository.getProfileById(profile.id ?: "") != null) {
                repository.updateProfile(profile)
            } else {
                repository.addProfile(profile)
            }
        }
    }

    fun defaultProfile(): ServerProfile? {
        val defaultProfile = repository.getDefaultProfile()
        if (defaultProfile != null) {
            return defaultProfile
        }

        // Fallback: check SharedPreferences for backward compatibility
        val defaultProfileId = sharedPreferences.getString(KEY_DEFAULT_PROFILE, null)
        if (defaultProfileId != null) {
            val profile = repository.getProfileById(defaultProfileId)
            if (profile != null) {
                // Migrate the default setting to Room
                repository.setDefaultProfile(defaultProfileId)
                sharedPreferences.edit().remove(KEY_DEFAULT_PROFILE).apply()
                return profile
            }
        }

        // If no default profile is set, return the first profile if available
        val profiles = profiles
        return if (profiles.isNotEmpty()) {
            profiles[0]
        } else null
    }

    fun setDefaultProfile(profile: ServerProfile) {
        profile.id?.let { repository.setDefaultProfile(it) }
    }

    fun addProfile(profile: ServerProfile) {
        if (profile.id.isNullOrEmpty()) {
            profile.id = UUID.randomUUID().toString()
        }

        // Set default service type if not set
        if (profile.serviceType.isNullOrEmpty()) {
            profile.serviceType = ServerProfile.TYPE_METUBE
        }

        repository.addProfile(profile)
    }

    fun updateProfile(profile: ServerProfile) {
        repository.updateProfile(profile)
    }

    fun deleteProfile(profile: ServerProfile) {
        repository.deleteProfile(profile)
    }

    fun hasProfiles(): Boolean {
        return repository.hasProfiles()
    }

    /**
     * Get profiles filtered by service type
     */
    fun getProfilesByServiceType(serviceType: String): List<ServerProfile> {
        return repository.getProfilesByServiceType(serviceType)
    }

    /**
     * Get all unique service types from existing profiles
     */
    fun allServiceTypes(): List<String> {
        val allProfiles = profiles
        val serviceTypes: MutableList<String> = ArrayList()

        for (profile in allProfiles) {
            val serviceType = profile.serviceType
            if (serviceType != null && !serviceTypes.contains(serviceType)) {
                serviceTypes.add(serviceType)
            }
        }

        return serviceTypes
    }

    /**
     * Get all torrent client types from existing profiles
     */
    fun allTorrentClientTypes(): List<String> {
        val allProfiles = profiles
        val clientTypes: MutableList<String> = ArrayList()

        for (profile in allProfiles) {
            if (ServerProfile.TYPE_TORRENT == profile.serviceType) {
                val clientType = profile.torrentClientType
                if (clientType != null && !clientTypes.contains(clientType)) {
                    clientTypes.add(clientType)
                }
            }
        }

        return clientTypes
    }

    companion object {
        private const val PREFS_NAME = "MeTubeSharePrefs"
        private const val KEY_PROFILES = "profiles"
        private const val KEY_DEFAULT_PROFILE = "default_profile"
    }
}